File: java/Taskie/app/src/main/java/ada/osc/taskie/view/TasksActivity.java
package ada.osc.taskie.view;

import android.content.Intent;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.widget.Toast;

import java.util.List;

import ada.osc.taskie.R;
import ada.osc.taskie.TaskRepository;
import ada.osc.taskie.model.Task;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class TasksActivity extends AppCompatActivity {

	private static final String TAG = TasksActivity.class.getSimpleName();
	private static final int REQUEST_NEW_TASK = 10;
	public static final String EXTRA_TASK = "task";

	TaskRepository mRepository = TaskRepository.getInstance();
	TaskAdapter mTaskAdapter;

	@BindView(R.id.fab_tasks_addNew) FloatingActionButton mNewTask;
	@BindView(R.id.recycler_tasks) RecyclerView mTasksRecycler;

	TaskClickListener mListener = new TaskClickListener() {
		@Override
		public void onClick(Task task) {
			toastTask(task);
		}

		@Override
		public void onLongClick(Task task) {
			mRepository.removeTask(task);
			updateTasksDisplay();
		}
	};

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_tasks);

		ButterKnife.bind(this);
		setUpRecyclerView();
		updateTasksDisplay();
	}

	private void setUpRecyclerView() {

		int orientation = LinearLayoutManager.VERTICAL;

		RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(
				this,
				orientation,
				false
		);

		RecyclerView.ItemDecoration decoration =
				new DividerItemDecoration(this, orientation);

		RecyclerView.ItemAnimator animator = new DefaultItemAnimator();

		mTaskAdapter = new TaskAdapter(mListener);

		mTasksRecycler.setLayoutManager(layoutManager);
		mTasksRecycler.addItemDecoration(decoration);
		mTasksRecycler.setItemAnimator(animator);
		mTasksRecycler.setAdapter(mTaskAdapter);
	}

	private void updateTasksDisplay() {
		List<Task> tasks = mRepository.getTasks();
		mTaskAdapter.updateTasks(tasks);
		for (Task t : tasks){
			Log.d(TAG, t.getTitle());
		}
	}

	private void toastTask(Task task) {
		Toast.makeText(
				this,
				task.getTitle() + "\n" + task.getDescription(),
				Toast.LENGTH_SHORT
		).show();
	}

	@OnClick(R.id.fab_tasks_addNew)
	public void startNewTaskActivity(){
		Intent newTask = new Intent();
		newTask.setClass(this, NewTaskActivity.class);
		startActivityForResult(newTask, REQUEST_NEW_TASK);
	}

	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {

		if(requestCode == REQUEST_NEW_TASK && resultCode == RESULT_OK) {
			if (data != null && data.hasExtra(EXTRA_TASK)) {
				Task task = (Task) data.getSerializableExtra(EXTRA_TASK);
				mRepository.saveTask(task);
				updateTasksDisplay();
			}
		}
	}
}

File: kotlin/Taskie/app/src/main/java/ada/osc/taskie/view/TasksActivity.kt
package ada.osc.taskie.view

import ada.osc.taskie.R
import ada.osc.taskie.common.getIntent
import ada.osc.taskie.common.toast
import ada.osc.taskie.model.Task
import ada.osc.taskie.persistance.TaskRepository
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import kotlinx.android.synthetic.main.activity_tasks.*

class TasksActivity : AppCompatActivity() {

    private val taskAdapter: TaskAdapter by lazy { TaskAdapter(taskClickListener) }

    private val taskClickListener: TaskClickListener = object : TaskClickListener {
        override fun onClick(task: Task) = showTaskMessage(task)

        override fun onLongClick(task: Task) {
            TaskRepository.removeTask(task)
            updateTasks()
        }
    }

    companion object {
        private val TAG = TasksActivity::class.java.simpleName
        private const val REQUEST_NEW_TASK = 10

        const val EXTRA_TASK = "task"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_tasks)

        initUi()
        updateTasks()
    }

    private fun initUi() {
        setUpRecyclerView()

        addNewTask.setOnClickListener { startActivityForResult(getIntent<NewTaskActivity>(), REQUEST_NEW_TASK) }
    }

    private fun setUpRecyclerView() {
        val orientation = LinearLayoutManager.VERTICAL

        val layoutManager = LinearLayoutManager(this, orientation, false)
        val decoration = DividerItemDecoration(this, orientation)
        val animator = DefaultItemAnimator()

        allTasks.apply {
            setLayoutManager(layoutManager)
            addItemDecoration(decoration)
            itemAnimator = animator
            adapter = taskAdapter
        }
    }

    private fun updateTasks() {
        val tasks = TaskRepository.getTasks()
        taskAdapter.updateTasks(tasks)

        tasks.forEach { task -> Log.d(TAG, task.title) }
    }

    private fun showTaskMessage(task: Task) = toast(task.title + "\n" + task.description)

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_NEW_TASK && resultCode == Activity.RESULT_OK) {
            if (data?.hasExtra(EXTRA_TASK) == true) {
                val task = data.getSerializableExtra(EXTRA_TASK) as Task
                TaskRepository.saveTask(task)
                updateTasks()
            }
        }
    }
}
